File: src/main/java/ru/xpendence/topjavagraduation/repository/RestaurantRepository.java
package ru.xpendence.topjavagraduation.repository;

import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import ru.xpendence.topjavagraduation.entity.Restaurant;

import java.time.LocalDate;
import java.util.List;
import java.util.Optional;

public interface RestaurantRepository extends JpaRepository<Restaurant, Long> {

    @Query("select r from Restaurant r left join Dish d on d.restaurant.id = r.id where d.id = ?1")
    Optional<Restaurant> getByDishId(Long dishId);

    @Query("""
            select r from Restaurant r
            join Vote v on v.restaurant.id = r.id
            where v.date = :date
            group by r.id
            order by count(v.id) desc
            """)
    List<Restaurant> findChosenByVoteDate(@Param("date") LocalDate date, Pageable pageable);

    void deleteById(Long id);
}

File: src/main/kotlin/ru/xpendence/topjavagraduation/repository/RestaurantRepository.kt
package ru.xpendence.topjavagraduation.repository

import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import ru.xpendence.topjavagraduation.entity.Restaurant
import java.time.LocalDate
import java.util.*

/**
 * Доступ к ресторанам через JDBC.
 */
interface RestaurantRepository {

    /**
     * Сохраняет ресторан: вставляет новую запись или обновляет имя существующей по [Restaurant.id].
     *
     * @param restaurant ресторан; для вставки — без id
     * @return сохранённый ресторан (с присвоенным id при вставке)
     */
    fun save(restaurant: Restaurant): Restaurant

    /**
     * Возвращает ресторан по идентификатору с блюдами и голосами за текущую дату.
     *
     * @param id идентификатор ресторана
     * @return ресторан с коллекциями или empty, если не найден
     */
    fun findById(id: Long): Optional<Restaurant>

    /**
     * Возвращает ресторан по идентификатору блюда с блюдами и голосами за текущую дату.
     *
     * @param dishId идентификатор блюда
     * @return ресторан или empty, если блюдо/ресторан не найдены
     */
    fun getByDishId(dishId: Long): Optional<Restaurant>

    /**
     * Возвращает страницу ресторанов с блюдами и голосами за текущую дату.
     *
     * @param pageable параметры страницы
     * @return страница ресторанов
     */
    fun findAll(pageable: Pageable): Page<Restaurant>

    /**
     * Возвращает рестораны, отсортированные по числу голосов на указанную дату (убывание).
     *
     * @param date дата голосования
     * @param pageable параметры страницы (обычно размер 1 — победитель дня)
     * @return список ресторанов с блюдами и голосами за текущую дату; пустой, если голосов нет
     */
    fun findChosenByVoteDate(date: LocalDate, pageable: Pageable): List<Restaurant>

    /**
     * Удаляет ресторан по идентификатору.
     *
     * @param id идентификатор ресторана
     */
    fun deleteById(id: Long)
}
